src/storage/txn/commands/flush.rs

```diff
@@ -0,0 +1,187 @@
+                    bytes += value.len();
+                }
+                Mutation::Delete(ref key, _) | Mutation::Lock(ref key, _) => {
+                    bytes += key.as_encoded().len();
+                }
+                Mutation::CheckNotExists(..) => (),
+            }
+        }
+        bytes
+    }
+
+    gen_lock!(mutations: multiple(|x| x.key()));
+}
+
+impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for Flush {
+    fn process_write(mut self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
+        let rows = self.mutations.len();
+        let mut txn = MvccTxn::new(self.start_ts, context.concurrency_manager);
+        let mut reader = ReaderWithStats::new(
+            SnapshotReader::new_with_ctx(self.start_ts, snapshot, &self.ctx),
+            context.statistics,
+        );
+        let mut old_values = Default::default();
+
+        let res = self.flush(&mut txn, &mut reader, &mut old_values, context.extra_op);
+        let locks = res?;
+        let extra = TxnExtra {
+            old_values,
+            one_pc: false,
+            allowed_in_flashback: false,
+        };
+        let new_locks = txn.take_new_locks();
+        let guards = txn.take_guards();
+        assert!(guards.is_empty());
+        Ok(WriteResult {
+            ctx: self.ctx,
+            to_be_write: WriteData::new(txn.into_modifies(), extra),
+            rows,
+            pr: ProcessResult::MultiRes { results: locks },
+            lock_info: vec![],
+            released_locks: ReleasedLocks::new(),
+            new_acquired_locks: new_locks,
+            lock_guards: guards,
+            response_policy: ResponsePolicy::OnApplied,
+            known_txn_status: vec![],
+        })
+    }
+}
+
+impl Flush {
+    fn flush(
+        &mut self,
+        txn: &mut MvccTxn,
+        reader: &mut SnapshotReader<impl Snapshot>,
+        old_values: &mut OldValues,
+        extra_op: ExtraOp,
+    ) -> Result<Vec<std::result::Result<(), crate::storage::errors::Error>>> {
+        let props = TransactionProperties {
+            start_ts: self.start_ts,
+            kind: TransactionKind::Optimistic(false),
+            commit_kind: CommitKind::TwoPc,
+            primary: &self.primary,
+            txn_size: 0, // txn_size is unknown
+            lock_ttl: self.lock_ttl,
+            min_commit_ts: TimeStamp::zero(),
+            need_old_value: extra_op == ExtraOp::ReadOldValue, // FIXME?
+            is_retry_request: self.ctx.is_retry_request,
+            assertion_level: self.assertion_level,
+            txn_source: self.ctx.get_txn_source(),
+        };
+        let mut locks = Vec::new();
+        // If there are other errors, return other error prior to `AssertionFailed`.
+        let mut assertion_failure = None;
+
+        for m in mem::take(&mut self.mutations) {
+            let key = m.key().clone();
+            let mutation_type = m.mutation_type();
+            let prewrite_result = prewrite(
```

Comment:
Some keys may be flushed twice with different values, this case is not handled by now, right?

Reply:
Yes

Reply:
What would happen if a key is prewritten twice by now?

Reply:
Write conflict error will be returned.

Reply:
The case is covered in `test_flush_overwrite`. 
The second prewrite will have no effects and results in an empty request. 
https://github.com/tikv/tikv/blob/e4e273f758c289df9ddf47b73371185bf867b2cd/src/storage/txn/actions/prewrite.rs#L80-L82

An `EmptyRequest` error is then returned when the engine tries to write it.